File: 03.programmers/2018_summer_winter_coding/01.C++/skill_tree.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

int solution(string skill, vector<string> skill_trees) {
	int answer = 0;
	vector<bool> check(26);    

	for (int i = 0; i < skill.length(); i++)    
		check[skill[i] - 65] = true;

	for (int i = 0; i < skill_trees.size(); i++) {
		int index = 0;
		int j;
        string s = skill_trees[i];
		for (j = 0; j < s.length(); j++) {
			if (check[s[j] - 65]) {
				if (skill[index] == s[j]) index++;
				else break;
			}
		}
		if (j == skill_trees[i].length())
			answer++;
	}
	return answer;
}
int main(){
    string skill = "CBD";
    vector<string> skill_trees = {"BACDE", "CBADF", "AECB", "BDA"};
    cout<<solution(skill,skill_trees)<<endl;
    return 0;
}

File: ~2023/03.programmers/2018_summer_winter_coding/02.Python/skill_tree.py
def solution(skill, skill_trees):
    answer = 0

    for skills in skill_trees:
        skill_list = list(skill)

        for s in skills:
            if s in skill:
                if s != skill_list.pop(0):
                    break
        else:
            answer += 1

    return answer


input_skill = "CBD"
input_skill_tree = ["BACDE", "CBADF", "AECB", "BDA"]
print(solution(input_skill,input_skill_tree))